qualtran/bloqs/generalized_qsp.py

```diff
@@ -79,6 +79,111 @@ def decompose_from_registers(
+
+    References:
+        [Generalized Quantum Signal Processing](https://arxiv.org/abs/2308.01501)
+            Motlagh and Wiebe. (2023). Theorem 4.
+    """
+    d = len(P) - 1  # degree
+
+    # R(z) = z^d (1 - P^*(z) P(z))
+    # obtained on simplifying Eq. 34, Eq. 35 by substituting H, T.
+    # The definition of $R$ is given in the text from Eq.34 - Eq. 35,
+    # following the chain of definitions below:
+    #
+    #     $$
+    #     T(\theta) = \abs{P(e^{i\theta}),
+    #     H = I - T,
+    #     H = e^{-id\theta} R(e^{i\theta})
+    #     $$
+    #
+    # Here H and T are defined on reals, so the initial definition of R is only on the unit circle.
+    # We analytically continue this definition to the entire complex plane by replacing $e^{i\theta}$ by $z$.
+    R = Polynomial.basis(d) - Polynomial(P) * Polynomial(np.conj(P[::-1]))
+    roots = R.roots()
+
+    # R is self-inversive, so larger_roots and smaller_roots occur in conjugate pairs.
+    units: list[complex] = []  # roots r s.t. \abs{r} = 1
+    larger_roots: list[complex] = []  # roots r s.t. \abs{r} > 1
+    smaller_roots: list[complex] = []  # roots r s.t. \abs{r} < 1
+
+    for r in roots:
+        if np.allclose(np.abs(r), 1):
+            units.append(r)
+        elif np.abs(r) > 1:
+            larger_roots.append(r)
+        else:
+            smaller_roots.append(r)
+
+    if verify:
+        # verify that the non-unit roots indeed occur in conjugate pairs.
+        def is_permutation(A, B):
+            assert len(A) == len(B)
+            A = list(A)
+            for z in B:
+                for w in A:
+                    if np.allclose(z, w):
+                        A.remove(w)
+                        break
+                else:
+                    return False
+            return True
+
+        assert is_permutation(smaller_roots, 1 / np.array(larger_roots).conj())
+
+    # Leading co-efficient of R described in Eq. 37.
+    c = R.coef[-1]
+    scaling_factor = np.sqrt(np.abs(c * np.prod(larger_roots)))
+
+    # pair up roots in `units`, claimed in Eq. 40 and the explanation preceding it.
+    # all unit roots must have even multiplicity.
+    paired_units: list[complex] = []
+    unpaired_units: list[complex] = []
+    for z in units:
+        matched_z = None
+        for w in unpaired_units:
+            if np.allclose(z, w):
+                matched_z = w
+                break
+
+        if matched_z:
+            paired_units.append(z)
+            unpaired_units.remove(matched_z)
+        else:
+            unpaired_units.append(z)
+
+    if verify:
+        assert len(unpaired_units) == 0
+
+    # Q = G \hat{G}, where
+    # - \hat{G}^2 is the monomials which are unit roots of R, which occur in pairs.
+    # - G*(z) G(z) is the interpolation of the conjugate paired non-unit roots of R,
+    #   described in Eq. 37 - Eq. 38
+    Q = scaling_factor * Polynomial.fromroots(paired_units + smaller_roots)
```

Comment:
Maybe a trivial question - Is my understanding correct that in the paper they seem to define `Q` with the opposite usage of `smaller_roots` and `larger_roots` ?  Something like

```python
scaling_factor = np.sqrt(np.abs(c * np.prod(smaller_roots)))
Q = scaling_factor * Polynomial.fromroots(paired_units + larger_roots)
```

This follows from Eq 37
<img width="1061" alt="image" src="https://github.com/quantumlib/Qualtran/assets/7863287/a34141bf-bc40-4fc1-8d5e-aa8d4d530f41">
where the scaling factor is 

$$
G = \sqrt{\prod_{k=1}^{m}\frac{1}{w_{k}^*}} \prod_{k=1}^{m} ||x - w_{k}||
$$

It would probably not matter but just want to make sure that I'm understanding this correct.

Reply:
Ah yes, that is true. I was debugging the issue of the polynomial being too large when substituting its zeroes, so I swapped the two around hoping it'll reduce the error.  I just tried swapping the two around, and it fails for degree 4.